### PR TITLE
--extra-*-dirs were not being passed to all components

### LINKED ISSUE #4720: cabal.project:extra-include-dirs not propagated to test-suites
Consider the repro-case at https://github.com/hvr/cabal-bugs/tree/master/issue-4720

which has in its `cabal.project` file the following directive

```
package bug
  -- orthogonal issue: base-dir ought to be cabal.project location,
  -- or there should be a way to refer to the project-folder, e.g.
  -- extra-include-dirs: $project_top/extra_inc

  extra-include-dirs: ../extra-inc
```

```
$ cd issue-4720/
$ cabal new-build :pkg:bug
Resolving dependencies...
Build profile: -w ghc-8.2.1 -O1
In order, the following will be built (use -v for more details):
 - bug-0 (lib) (first run)
 - bug-0 (exe:bug-exe1) (first run)
Configuring library for bug-0..
Configuring executable 'bug-exe1' for bug-0..
Warning: 'include-dirs: ../extra-inc' is a relative path outside of the source
tree. This will not work when generating a tarball with 'sdist'.
Warning: 'include-dirs: ../extra-inc' is a relative path outside of the source
tree. This will not work when generating a tarball with 'sdist'.
Warning: 'include-dirs: ../extra-inc' is a relative path outside of the source
tree. This will not work when generating a tarball with 'sdist'.
Warning: 'include-dirs: ../extra-inc' is a relative path outside of the source
tree. This will not work when generating a tarball with 'sdist'.
Preprocessing library for bug-0..
Building library for bug-0..
Preprocessing executable 'bug-exe1' for bug-0..
Building executable 'bug-exe1' for bug-0..
[1 of 1] Compiling LibBug           ( LibBug.hs, /stuff3/work/GitHub/cabal-bugs/issue-4720/dist-newstyle/build/x86_64-linux/ghc-8.2.1/bug-0/build/LibBug.o )
[1 of 2] Compiling LibBug           ( LibBug.hs, /stuff3/work/GitHub/cabal-bugs/issue-4720/dist-newstyle/build/x86_64-linux/ghc-8.2.1/bug-0/c/bug-exe1/build/bug-exe1/bug-exe1-tmp/LibBug.o )
[2 of 2] Compiling Main             ( Main.hs, /stuff3/work/GitHub/cabal-bugs/issue-4720/dist-newstyle/build/x86_64-linux/ghc-8.2.1/bug-0/c/bug-exe1/build/bug-exe1/bug-exe1-tmp/Main.o )

In file included from cbits/bug.c:4:0: error: 

../extra-inc/bug-extra-inc.h:2:2: error:
     warning: #warning yay [-Wcpp]
     #warning yay
      ^
  |
2 | #warning yay
  |  ^

In file included from cbits/bug.c:4:0: error: 

../extra-inc/bug-extra-inc.h:2:2: error:
     warning: #warning yay [-Wcpp]
     #warning yay
      ^
  |
2 | #warning yay
  |  ^

In file included from cbits/bug.c:4:0: error: 

../extra-inc/bug-extra-inc.h:2:2: error:
     warning: #warning yay [-Wcpp]
     #warning yay
      ^
  |
2 | #warning yay
  |  ^
Linking /stuff3/work/GitHub/cabal-bugs/issue-4720/dist-newstyle/build/x86_64-linux/ghc-8.2.1/bug-0/c/bug-exe1/build/bug-exe1/bug-exe1 ...

```

this works as expected, but if you enable the test-suite, this fails unexpectedly:

```
$ cabal new-build --enable-tests :pkg:bug
Resolving dependencies...
Build profile: -w ghc-8.2.1 -O1
In order, the following will be built (use -v for more details):
 - bug-0 (lib) (configuration changed)
 - bug-0 (exe:bug-exe1) (configuration changed)
 - bug-0 (test:bug-test1) (first run)
Configuring library for bug-0..
Configuring executable 'bug-exe1' for bug-0..
Configuring test suite 'bug-test1' for bug-0..
Warning: 'include-dirs: ../extra-inc' is a relative path outside of the source
tree. This will not work when generating a tarball with 'sdist'.
Warning: 'include-dirs: ../extra-inc' is a relative path outside of the source
tree. This will not work when generating a tarball with 'sdist'.
Warning: 'include-dirs: ../extra-inc' is a relative path outside of the source
tree. This will not work when generating a tarball with 'sdist'.
Warning: 'include-dirs: ../extra-inc' is a relative path outside of the source
tree. This will not work when generating a tarball with 'sdist'.
Warning: 'include-dirs: ../extra-inc' is a relative path outside of the source
tree. This will not work when generating a tarball with 'sdist'.
Warning: 'include-dirs: ../extra-inc' is a relative path outside of the source
tree. This will not work when generating a tarball with 'sdist'.
Preprocessing library for bug-0..
Building library for bug-0..
Preprocessing executable 'bug-exe1' for bug-0..
Building executable 'bug-exe1' for bug-0..
Preprocessing test suite 'bug-test1' for bug-0..
Building test suite 'bug-test1' for bug-0..
[1 of 2] Compiling LibBug           ( LibBug.hs, /stuff3/work/GitHub/cabal-bugs/issue-4720/dist-newstyle/build/x86_64-linux/ghc-8.2.1/bug-0/c/bug-test1/build/bug-test1/bug-test1-tmp/LibBug.o )
[2 of 2] Compiling Main             ( Main.hs, /stuff3/work/GitHub/cabal-bugs/issue-4720/dist-newstyle/build/x86_64-linux/ghc-8.2.1/bug-0/c/bug-test1/build/bug-test1/bug-test1-tmp/Main.o )

cbits/bug.c:4:27: error:
     fatal error: bug-extra-inc.h: No such file or directory
  |
4 | #include <bug-extra-inc.h>
  |                           ^
compilation terminated.
`gcc' failed in phase `C Compiler'. (Exit code: 1)
cabal: Failed to build test:bug-test1 from bug-0.
```

IOW, the `extra-include-dirs` were not passed properly to the compilation of the test-suites cbits

### PR DESCRIPTION
**Pull request by Ericson2314**
Only to libraries and executables before this.

Fixes #4720

Please include the following checklist in your PR:

* [ ] Patches conform to the [coding conventions](https://github.com/haskell/cabal/#conventions).
* [ ] Any changes that could be relevant to users have been recorded in the changelog.
* [ ] The documentation has been updated, if necessary.
* [ ] If the change is docs-only, `[ci skip]` is used to avoid triggering the build bots.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

**Comment by Ericson2314** [comment]
Maybe in 6040df41d4b17ae6694f1f761b748e1b31a9bb65 those were the only two types of components? This code's been wrong for a while.

I ran into this bootstrapping GHC, so I'd like to do a number of back-ports, if possible.

**Comment by hvr** [comment]
@Ericson2314 iirc I looked up when benchmark components were added... and I found out it was in Cabal 1.9.2; so your theory sounds quite likely

**Comment by 23Skidoo** [comment]
Code looks fine. What additional testing besides CI do you have in mind?

**Comment by Ericson2314** [comment]
@23Skidoo https://github.com/hvr/cabal-bugs/tree/master/issue-4720 perhaps would be a good one? Don't tempt me to just merge, do the backports, and be on my way without improving the testsuite! :D

**Comment by 23Skidoo** [comment]
A test case would be appreciated of course!

File: Cabal/Distribution/Simple/Configure.hs [hunk 1]
```diff
@@ -966,14 +966,24 @@ configureFinalizedPackage verbosity cfg enabled
...
                              , PD.includeDirs = configExtraIncludeDirs cfg}
-            modifyLib l        = l{ libBuildInfo = libBuildInfo l
-                                                   `mappend` extraBi }
-            modifyExecutable e = e{ buildInfo    = buildInfo e
-                                                   `mappend` extraBi}
-        in pkg_descr{ library = modifyLib `fmap` library pkg_descr
-                    , subLibraries = modifyLib `map` subLibraries pkg_descr
-                    , executables = modifyExecutable  `map`
-                                      executables pkg_descr}
+            modifyLib l        = l{ libBuildInfo        = libBuildInfo l
```
Comment (hvr) [inline Cabal/Distribution/Simple/Configure.hs]: ...I'd just point out that this would a good opportunity for leveraging lens vocabulary... :-)

**Comment by 23Skidoo** [comment]
I think I'll merge this to `2.0`, but leave the PR for `master` open for now to gently nudge @Ericson2314 into writing a test case.

**Comment by hvr** [comment]
just for the record, this was already merged to 2.0 via f7f7fa07326ddc6ffb954b4ece92369d3447fa9c

**Comment by 23Skidoo** [comment]
@Ericson2314 Ping.

**Comment by kmicklas** [comment]
@hvr I added a minimized test case based on your repo for @Ericson2314.

**Comment by hvr** [comment]
@23Skidoo ping

**Comment by Ericson2314** [comment]
(And for the record, previously backported in 6394cb0b6eba91a8692a3d04b2b56935aed7cccd too, again with test.)

**Comment by 23Skidoo** [comment]
@kmicklas Thanks! Looks like this can go in now.

**Pull request merged**